CONTRIBUTING.md

```diff
@@ -141,19 +141,22 @@ tracing) enabled.
 
 The setup for a local rustc works as follows:
 ```sh
+# Clone the rust-lang/rust repo.
 git clone https://github.com/rust-lang/rust/ rustc
 cd rustc
 cp config.toml.example config.toml
 # Now edit `config.toml` and set `debug-assertions = true`.
-# This step can take 30 minutes and more.
-./x.py build src/rustc
+
+# Build a stage 2 rustc.
+# This step can take 30 minutes or more.
+./x.py build compiler/rustc --stage 2
 # If you change something, you can get a faster rebuild by doing
-./x.py --keep-stage 0 build src/rustc
+./x.py --keep-stage 0 build compiler/rustc --stage 2
```

Comment:
Also didn't check if `--keep-stage 0` still works.

Reply:
It shouldn't have been affected by the x.py changes. But `compiler/rustc` is not necessary here either.
```suggestion
./x.py --keep-stage 0 --stage 2
```

Reply:
But won't it build rustdoc etc. if you leave out `compiler/rustc`?

Also: I'm guessing you didn't mean to suggest removing `build` ;)

Reply:
Sure, if you don't want to build tools you could add a filter:
```suggestion
./x.py build --keep-stage 0 --stage 2 library/std
```
I'm not sure how miri works - is miri normally built in-tree or out of tree? If it's in tree you'll also need `src/tools/miri` on the end.

Reply:
I *think* it's built out of tree, but I'm not sure.

And won't `library/std` only build `std` and not the compiler? That or I don't understand bootstrap filters at all, which is a possibility ;)

Reply:
library/std builds std _with_ the new compiler. I recommend reading through https://rustc-dev-guide.rust-lang.org/building/bootstrapping.html#stages-and-std.

Reply:
Okay, I just read that, and I *think* I understand now: `--stage 2` means:

1. Compile std with beta rustc
2. Compile rustc with beta rustc and std from step 1
3. Compile std with rustc from step 2
4. Compile rustc with rustc from step 2 and std from step 3

Is that correct?

Reply:
> `./x.py build --stage 2 library/std`

That's what `build --stage 1 compiler/rustc` does. `build --stage 2 library/std` has a 5th step, 'compile std with rustc from step 4'.

> ./x.py build --keep-stage 0 --stage 2 library/std

No, `--keep-stage` means keep the cached artifacts from before (steps 3 and 4) and assume they don't have to be changed. Then run step 5 as before.

Reply:
Okay, I still feel confused. A bit less confused, but I think I will have to do some deep thinking on this ;)

I'll apply your patches since you're the `x.py` guru.

Reply:
If it helps, 'previous compile' is referring to the last time you _ran_ x.py, not the previous stage.

Reply:
Huh, okay.

Reply:
FWIW, the old command did exactly what is needed and no more. :) So you'll have to keep pointing at `compiler/rustc`. Only `x.py` defaults changed, not the actual behavior, so why would y'all think that a different filter is needed now?

We need the actual stage 2 compiler, not just a standard library for stage 2. That's why it is `--stage 2 compiler/rustc`.